library/core/src/intrinsics.rs

```diff
@@ -2385,6 +2385,32 @@ extern "rust-intrinsic" {
     #[rustc_nounwind]
     pub fn raw_eq<T>(a: &T, b: &T) -> bool;
 
+    /// Lexicographically compare `[left, left + bytes)` and `[right, right + bytes)`
+    /// as unsigned bytes, returning negative if `left` is less, zero if all the
+    /// bytes match, or positive if `right` is greater.
+    ///
+    /// This underlies things like `<[u8]>::cmp`, and will usually lower to `memcmp`.
+    ///
+    /// # Safety
+    ///
+    /// * If `bytes` is zero, `left` and `right` must be non-null.
+    ///
+    /// * If `bytes` is non-zero,
+    ///
+    ///   * `left` must be [valid] for reads of `bytes` bytes,
+    ///     all of which must be initialized.
+    ///
+    ///   * `right` must be [valid] for reads of `bytes` bytes,
+    ///     all of which must be initialized.
```

Comment:
I think this should follow the usual documentation of all our intrinsics: left and right must be valid for reads of size `bytes`. That's all we need to say.

We already say that as far as Rust is concerned, `4 as *const u8` is valid for reads of size 0.

Reply:
In particular, if we ever declare that even the null pointer is valid for zero-sized reads and writes (which we may, https://github.com/rust-lang/opsem-team/issues/12), then even passing null to these intrinsics should be allowed.